### PR TITLE
Detection modularity

### PR DESCRIPTION
**Pull request by andreitudorp**

### CODE REVIEW

File: utils/result.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        self.url = url
+        self.url_hash = url_hash
+        self.status = status
+        self.result = result
+
+    def to_json_str(self):
+        # TODO rename status to phase (over the whole codebase)
+        obj = [
+            {"url": self.url, "status": self.status, "result": self.result.name, "hash": self.url_hash}
+        ]
```
Comment (TPGamesNL) [inline utils/result.py:42]: why is this an array?

File: utils/result.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class ResultTypes(Enum):
+    PROCESSING = -2
+    LEGITIMATE = -1
+    INCONCLUSIVE = 0
+    PHISHING = 1
+
+    @classmethod
+    def to_old(cls, result: "ResultTypes"):
```
[the remark this answers is not in the document]
Comment (TPGamesNL) [reply utils/result.py:12]: also, should this func have a deprecation notice?

File: utils/decision.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+from enum import Enum, auto
+
+from utils.result import ResultTypes
+
+
+class DecisionStrategies(Enum):
+    Strict = auto()
+    Majority = auto()
+    Unanimous = auto()
```
Comment (TPGamesNL) [inline utils/decision.py:9]: arent enum supposed to be screaming snake case

File: utils/decision.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def decide(strategy: DecisionStrategies, results: list[ResultTypes]) -> ResultTypes:
+    """
+    Given a decision strategy and a list of results, it computes the decision.
+    """
+
+    match strategy:
+        case DecisionStrategies.Strict:
```
Comment (TPGamesNL) [inline utils/decision.py:18]: is there a thing in python where you can attach a value (like a function) to an enum const? then we can add the _strict function etc to it and use it directly

File: utils/decision.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+from enum import Enum, auto
+
+from utils.result import ResultTypes
+
+
+class DecisionStrategies(Enum):
```
Comment (TPGamesNL) [inline utils/decision.py:6]: each decision should have documentation saying what it does. either in their funky impls or the enum values, but e.g. for strict: 'computes the strict decision' doesnt explain what it does

File: utils/decision.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    Given a list of results it computes the unanimous decision.
+    """
+
+    length = len(results) - results.count(ResultTypes.INCONCLUSIVE)
+    diff = results.count(ResultTypes.PHISHING) - results.count(ResultTypes.LEGITIMATE)
+
+    if abs(diff) != length:
+        return ResultTypes.INCONCLUSIVE
+
+    return ResultTypes(max(-1, min(1, diff)))
```
Comment (TPGamesNL) [inline utils/decision.py:60]: what exactly is this supposed to do for non-zero inconclusive results?

File: methods/detection_methods.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+class DetectionMethod(ABC):
+    @abstractmethod
+    def test(self) -> ResultTypes:
+        pass
+
+
+class TestMethod(DetectionMethod):
+    def test(self, *args, **kwargs):
+        rand = random.randint(-1, 1)
+        return ResultTypes(rand)
```
Comment (TPGamesNL) [inline methods/detection_methods.py:18]: either remove or document this

File: detection.py [hunk 8]
```diff
@@ -1,289 +1,133 @@
...
-async def check_search_results(url_registered_domain, found_urls) -> "DetectionResult":
-    with TimeIt("SAN domain check"):
-        with concurrent.futures.ThreadPoolExecutor() as pool:
-            loop = asyncio.get_running_loop()
-            coros = []
-            for url in found_urls:
-                coros.append(loop.run_in_executor(pool, lambda: check_url(url_registered_domain, url)))
+    for method in settings.detection_methods:
+        main_logger.info(f"Started running method {method.name}")
+        results.append(method.value.test(data.url, data.screenshot_url, data.uuid, data.pagetitle, None))
```
Comment (TPGamesNL) [inline detection.py:66]: didnt you want to remove image64?

File: methods/reverse_image_search_method.py [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+
+# The HTTP + HTML session to use for reverse image search
+html_session = HTMLSession()
+html_session.browser  # TODO why is this here
+
+# The logo classifier, deserialized from file
+logo_classifier = joblib.load("saved-classifiers/gridsearch_clf_rt_recall.joblib")
+
+
+class ReverseImageSearchMethod:
```
Comment (TPGamesNL) [inline methods/reverse_image_search_method.py:48]: doesnt extend the abc

File: utils/decision.py (hunk 4 above)
Comment (TPGamesNL) [reply utils/decision.py:18]: what if we switch to an abc and subclasses system instead? that way you also dont need the extra `decide` function cause it'd just be `strategy.decide(results)`

File: utils/decision.py (hunk 5 above)
Comment (TPGamesNL) [reply utils/decision.py:6]: also shouldnt this name be singular, `DecisionStrategy`?

File: detection.py [hunk 10]
```diff
@@ -1,289 +1,133 @@
...
+        self.decision_strategy = decision_strategy
+        self.bypass_cache = bypass_cache
 
+    @classmethod
+    def from_json(cls, json):
+        cls.detection_methods = [
+            DetectionMethods[method]
+            for method in json["detection-methods"]
+            if method in DetectionMethods.__members__
+        ]
```
Comment (TPGamesNL) [inline detection.py:95]: as discussed over vc i think it'd be better not to use reflection

File: utils/result.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+from enum import Enum
+
+
+class ResultTypes(Enum):
+    PROCESSING = -2
+    LEGITIMATE = -1
+    INCONCLUSIVE = 0
+    PHISHING = 1
+
+    @classmethod
```
Comment (TPGamesNL) [inline utils/result.py:11]: why is this a class method?

File: detection.py [hunk 12]
```diff
@@ -1,289 +1,133 @@
...
+        self,
+        detection_methods: list[DetectionMethods] = [DetectionMethods.ReverseImageSearch],
+        decision_strategy: DecisionStrategies = DecisionStrategies.Majority,
+        bypass_cache: bool = False,
+    ):
+        self.detection_methods = detection_methods
+        self.decision_strategy = decision_strategy
+        self.bypass_cache = bypass_cache
 
+    @classmethod
```
Comment (TPGamesNL) [inline detection.py:89]: wouldnt it be better to make this static instead? otherwise calling it with an instance will overwrite its data

Comment (TPGamesNL) [reply detection.py:89]: same for DetectionData

File: api.py [hunk 13]
```diff
@@ -31,28 +34,25 @@ def shutdown_server():
...
-    # extra json field for evaluation purposes
-    # the hash computed in the DB is the this one
-    if "phishURL" in json:  # TODO only allow this on a testing environment, not prod
-        url = json["phishURL"]
-        main_logger.info(f"Real URL changed to phishURL: {url}\n")
 
-    uuid = json["uuid"]
-    pagetitle = json["pagetitle"]
-    image64 = json["image64"]
+@app.route("/api/v1/url_new", methods=["POST"])
```
Comment (TPGamesNL) [inline api.py:47]: given that the api endpoint contains version numbers, shall we use them to make a new file for v1 api and switch this one to the main v2?

File: utils/result.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
+import json
+from enum import Enum
+
+
+class ResultTypes(Enum):
```
Comment (TPGamesNL) [inline utils/result.py:5]: shouldnt this name be singular, ResultType?

File: methods/__init__.py [hunk 15]
```diff
@@ -0,0 +1,9 @@
+from abc import ABC, abstractmethod
+
+from utils.result import ResultType
+
+
+class DetectionMethod(ABC):
+    @abstractmethod
+    def run(self) -> ResultType:
```
Comment (TPGamesNL) [inline methods/__init__.py:8]: arent there params missing here?

File: utils/decision.py [hunk 16]
```diff
@@ -1,60 +1,51 @@
...
-from utils.result import ResultTypes
+from utils.result import ResultType
 
 
-class DecisionStrategies(Enum):
-    Strict = auto()
-    Majority = auto()
-    Unanimous = auto()
+class DecisionStrategy(ABC):
+    @staticmethod
```
Comment (TPGamesNL) [inline utils/decision.py:7]: why is this one static but the other abc's things arent?

Comment (andreitudorp) [reply utils/decision.py:7]: What do you mean by other abc's stuff? It is static since it is not dependent on the instance of the DecisionStrategy

Comment (TPGamesNL) [reply utils/decision.py:7]: The other abstract classes from this PR, or are there not multiple? Don't they have the same property?

File: methods/__init__.py (hunk 15 above)
Comment (andreitudorp) [reply methods/__init__.py:8]: Yes and no, each method could have a different set of parameters. We could set it to the current one used by the dst, but most probably we would need to change it in the future.

Comment (TPGamesNL) [reply methods/__init__.py:8]: Aight just keep it for now then

File: utils/decision.py (hunk 16 above)
Comment (andreitudorp) [reply utils/decision.py:7]: image

This is the other abstract class that exists. I am not sure if this should also be abstract.

Comment (TPGamesNL) [reply utils/decision.py:7]: Yea fair enough, that one may have some use to instances

**Pull request merged**